### PR TITLE
perf: Implement faster sync methods for postgres

### LINKED ISSUE #48575: Faster sync on Postgres

### PR DESCRIPTION
**Pull request by snoe**
Fixes #48575

Pulls work from redshift into the common postgres driver.

### CODE REVIEW

File: src/metabase/driver/postgres.clj [hunk 1]
```diff
@@ -278,13 +279,105 @@
   ;; memory in a set like this
   {:tables (into #{} (describe-syncable-tables database))})
 
+(defmethod sql-jdbc.sync/describe-fields-sql :postgres
+  ;; The implementation is based on `getColumns` in https://github.com/aws/amazon-redshift-jdbc-driver/blob/master/src/main/java/com/amazon/redshift/jdbc/RedshiftDatabaseMetaData.java
+  ;; The `database-is-auto-increment` and `database-required` columns are currently missing because they are only
+  ;; needed for actions, which redshift doesn't support yet.
```
Comment (metamben) [inline src/metabase/driver/postgres.clj:285]: Why is it relevant here that the redshift driver doesn't support actions yet?

File: src/metabase/driver/postgres.clj [hunk 2]
```diff
@@ -278,13 +279,105 @@
...
-  (binding [*enum-types* (enum-types driver database)]
-    (sql-jdbc.sync/describe-table driver database table)))
+;; name, but first fetches database enum types so we have access to them.
+(defmethod driver/describe-fields :postgres
+  [driver database & args]
+  (let [enums (enum-types database)]
+    (eduction
+     (map (fn [{:keys [database-type] :as col}]
+            (cond-> col
+              (contains? enums (keyword database-type))
```
Comment (metamben) [inline src/metabase/driver/postgres.clj:378]: Is the conversion to keyword useful?

**Review by metamben — APPROVED** [review]
LGTM, but it's very difficult to tell how well it's going to work in the wild. Is there a way to test this against schemas from our customers' DWs?

**Comment by artem-levashov** [comment]
I have 6 differend DBs. all 6 has tables with exact same name. tables has different  fields. after scan DB table metadata of sunced table contains merged list of  fields from all 6 tables. it started after I upated to version 0.51.3   Fir 8 Nov 2024.  Is it possible that  toptimization from this  pull reuqest generates  my issue?

**Comment by artem-levashov** [comment]
image
loos like lost filter by  TABLE_SCHEMA

**Pull request merged**